### ISSUE TITLE
Create tap-hold-release button

### ISSUE THREAD

**Issue by david-janssen** [david-janssen/kmonad#48]
So, referencing the beginning of #38 and the end of #42, it seems there is continued interest in the following style of button.

A tap-hold style button combinator that decides if it enters 'held' or 'tapped' mode based on the next *release*. If the first release it detects is itself, then we act as a tap button. However, if the next release we detect is of another key, then we act as held (with rollback of any events that might have occurred).

This as opposed to:
- a tap-next: which decides on if the next event is the 'release of itself' or not
- a tap-hold: which decides on a delay (with rollback)
- a tap-hold-next: which is a mixture of the above two (further documented in #42)

@thoelze1 has expressed interest in trying his hand at this. I think I have an idea of how to do this but I'll let you give it a shot. As usual I am available for questions.

Just mentioning @obar as well, for notification reasons.

Trying to keep the issues organized :-)

EDIT: clarification

**Comment by obar** [comment]
I'm looking forward to it! I suspect that a version of this with a timeout (as `tap-hold-next` is to `tap-hold`) would also be useful.

For further cross-referencing, in QMK they do this using [ignore mod tap interrupt](https://docs.qmk.fm/#/tap_hold?id=ignore-mod-tap-interrupt), and another search term someone might use is permissive hold.

**Comment by david-janssen** [comment]
Thanks for the reference documentation, I am starting to see how this could be a very useful feature indeed.

I do want to say, I would really like to avoid using global settings to modify key behavior. I'd much rather have a new button type that performs as desired, that way each button can be specialized to its requirements, without affecting others. It is also much easier to reason about what a button does when it all happens in isolation.

**Comment by obar** [comment]
Absolutely, and I like your approach. The default to using globals on QMK makes sense an embedded system where memory is a valuable commodity, but that's not the case here.

**Comment by david-janssen** [comment]
Yeah, in embedded systems it makes loads of sense :-)

**Comment by thoelze1** [comment]
I think I've figured out a solution, working on implementation now

**Comment by david-janssen** [comment]
Sorry, didn't get a notification of this, so I didn't see it until now. Will have a look :-)

**Comment by david-janssen** [comment]
Had a look, I agree with everything conceptually. I wouldn't have thought to check specifically for buttons that were pre-pressed, so I wouldn't have included all the machinery for detecting *first* a press and *then* a release, however that is actually a good idea when it comes to robust buttons.

2 points:

1. The code in your function is a bit difficult to read (especially nested if statements need a lot of mental parsing). Also, I am not sure what the point of `f` is. So I've 'cleaned' it up (subjective, I am sure there are people out there who probably cringe at my compulsion to outline haskell code :-)) to this:

```haskell

-- | Create a 'Button' that performs a tap of some button if the next release of
-- any keypress not prior to this keypress comes from this keypress, or else
-- switches to holding some other button if the next release of any keypress not
-- prior to this keypress comes from another keypress.
tapNextRelease :: Button -> Button -> Button
tapNextRelease t h = onPress $ do
  hold True
  go []
  where

    doTap, doHold :: MonadK m => m Catch
    doTap  = tap t   *> hold False *> pure Catch
    doHold = press h *> hold False *> pure NoCatch
   
    go :: MonadK m => [Keycode] ->  m ()
    go ks = hookF $ \e -> do
      p <- matchMy Release
      let isRel = isRelease e
      if | p e                             -> doTap
         | isRel && (e^.keycode `elem` ks) -> doHold
         | not isRel                       -> go ((e^.keycode):ks) *> pure NoCatch
         | otherwise                       -> go ks *> pure NoCatch
```

2. This button description is a bit too complicated for my tastes. However, in this case it is (currently) completely warranted. The reason it is too complicated is because it implements a new behavior we haven't put anywhere before, which is the case of: "match against a particular pattern, once the entire pattern is satisfied do something". Ideally, we'd create a combinator that implements this "wait for a pattern" behavior, and then rewrite this button in terms of that combinator. That will make future "wait for a pattern" buttons much easier to write.

However, both of these comments are academic. Your code looks good! Have you tried it? I eagerly await your pull-request.

**Comment by thoelze1** [comment]
Thanks for all the feedback. I definitely like the "wait for a pattern" idea.

I forgot to mention in my last comment that the code I posted doesn't yet work. I just tried your edit and it still doesn't work. I don't yet have a strong grasp on monads, but I'll keep trying to debug.

**Comment by david-janssen** [comment]
Ahhh, I was being lazy :-). If you add the code for parsing and joining, then it'll be easy to test. Then I can join in the efforts tomorrow. Should we experiment on a new branch?

**Comment by thoelze1** [comment]
I just made a pull request. I used the `testing` branch instead of a new one because GitHub wouldn't let me PR from a branch that didn't yet exist upstream.

**Comment by david-janssen** [comment]
Merged into testing, not going to work on KMonad today, completely knackered.

**Comment by thoelze1** [comment]
This is the behavior I currently observe on `testing`. When I do the following:
- press a `tap-next-release`
- press a non-modifier key
- release the `tap-next-release`
- release the non-modifier key

I get the "hold" behavior for the `tap-next-release` even though I should get the "tap" behavior. Is it possible that this is because when a non-modifier key is pressed, a two key events (corresponding to press and release) are immediately issued when the physical key is pressed?

I also observe this when I do the following:
- press a non-modifier key
- press a `tap-next-release`
- press another non-modifier key
- release the first non-modifier key

Then I again get the "hold" behavior. Any ideas? Having a lot of trouble figuring this out.

**Comment by obar** [comment]
I just tried applying the rewrite that @david-janssen [proposed above](https://github.com/david-janssen/kmonad/issues/48#issuecomment-671655824) to the testing branch, and when trying your first example I get expected behavior, so it works for rolls:

> * press a `tap-next-release`
> * press a non-modifier key
> * release the `tap-next-release`
> * release the non-modifier key

The second example also works as expected, as long as the keys are released in the pressed order.

However when trying to use the `tap-next-release` as a modifier (release in a different order than pressed), the modified key ends up stuck in a pressed state with no release until the next event:

* press a `tap-next-release`
* press `g`, a non-modifier key
* release `g`
* release the `tap-next-release

I get `mod-g` followed by a string of gggggggggggg repeating until I press another key.

Steps in the right direction at any rate!

**Comment by david-janssen** [comment]
Hey guys,

Great to see you're making progress. Work's been crazy busy this week, but I hope to have a bit of time to look at all of this on Sunday. Can't wait to get stuck in :)

**Comment by david-janssen** [comment]
Looked at this in more depth, figured out the problem:

In the case where we use tap-next-release as a modifier we:
Press tap-next-release
Get the press of some other key and store it
Get the release of some other key, which triggers hold-mode. Then we rerun all the stored events.

That means we rerun the stored press after we've just handled its release, resulting in a state where we emit that press but never its matching release.

The simple solution is to then take the button that was tapped, remove it from the rerun-store, and just programatically tap it. This required a lot of manipulation of things that were never meant to be manipulated, and the code gets very messy. Going to take some time to think of a cleaner solution, because it feels like we're going against the grain a.t.m.

I think I will add a new, basic KMonad action: 'throwEvent' that allows for the insertion of an input event to the front of KMonad. That way we could opt to catch an event in the handler, do everything we want to do, and then rethrow the event, ensuring that it will happen after all of our changes. (Because a.t.m. the  issue is that the release triggers some stuff, including the press, but then can't be rethrown again).

**Comment by david-janssen** [comment]
Yo, got it working. Just as I described in my previous post, I had to add a new 'inject' primitive, to insert new input events into KMonad, so that we could handle events, catch them, and then rethrow them, to change up the ordering of how events are processed. I like to keep the primitives to a minimum, because that keeps the program easier to reason about, but in this case it turns out that adding the primitive was:
1) Very easy
2) Made the solution to our problem very easy.

Therefore, probably a good primitive :-).

Anyways, it's still on the 'testing' branch. Give it a go and let me know what you think. If everyone is happy I'll merge this into refactor soon. Also, I'm thinking we should merge refactor into master soon, since I have almost run out of issues, and I think refactor is probably more stable than master.

**Comment by obar** [comment]
I just ran through the same tests I've done before, this time with no issues whatsoever. The new button type also unsurprisingly composes the same way as `tap-hold-next` did for another useful mod I've got to turn on a number pad for the right hand: 

```
bnm (tap-next-release b (layer-toggle numbers))
```

A nice-to-have would be a variant with a timeout (`tap-next` : `tap-hold-next` :: `tap-next-release` : **`tap-hold-next-release`**) for the oops-I-shouldn't-have-pressed-this-mod situations, but it's not strictly necessary. I'm sure I will need it less as I get used to these new home row modifiers.

Thank you for this!

**Comment by david-janssen** [comment]
Good to hear it works for you. In regards to the timeout variant, I was actually just thinking the same. I'll have a go, although I think we should really do something about our naming-scheme. Perhaps have buttons with an optional timeout argument? Although I think I'll just implement this now and we can think about a refactor later.

**Comment by david-janssen** [comment]
Think I have the timeout variant working as well: `(tap-hold-next-release 200 a sft)`, for example.

**Comment by thoelze1** [comment]
I rebuilt from `testing` but still having problems. This is my config:

```
(defcfg
  input (iokit-name)
  output (kext)
)

(defalias
  spc (tap-next-release spc lsft)
)

(defsrc
  esc  f1   f2   f3   f4   f5   f6   f7   f8   f9   f10  f11  f12 
  grv  1    2    3    4    5    6    7    8    9    0    -    =    bspc                
  tab  q    w    e    r    t    y    u    i    o    p    [    ]    \
  caps a    s    d    f    g    h    j    k    l    ;    '    ret
  lsft z    x    c    v    b    n    m    ,    .    /    rsft up
  fn   lctl lalt lmet           spc            rmet ralt left down rght
)

(deflayer hyperspace
  esc  f1   f2   f3   f4   f5   f6   f7   f8   f9   f10  f11  f12
  grv  1    2    3    4    5    6    7    8    9    0    -    =    bspc                
  tab  q    w    e    r    t    y    u    i    o    p    [    ]    \
  caps a    s    d    f    g    h    j    k    l    ;   '    ret
  lsft z    x    c    v    b    n    m    ,    .    /    rsft up
  fn   lctl lalt lmet           @spc           rmet ralt left down rght
)
```

The `tap-next-release` never sends space. Am I missing something?

**Comment by obar** [comment]
@david-janssen Solid point on the naming convention. I need to test the timeout variant more thoroughly but thank you for implementing that as well!

@thoelze1 I just added your `spc` definition to my config and I am typing with it now. Works fine here. Are you sure your hyperspace layer is active? Maybe try swapping some letters to make it obvious.

It looks as though your alignment is off for your hyperspace layer, at `'` and `ret. I don't know if that would cause your problem though (would think not).

**Comment by david-janssen** [comment]
alignment shouldn't matter at all anymore. I can have a look at your config tomorrow. What are the exact issues you are having? Even just tapping `@spc` does nothing?

**Comment by thoelze1** [comment]
@obar The config I posted is simplified for easier testing - I currently have other buttons active in the layer and they work.

@david-janssen I downloaded a fresh copy of your repo and built from `testing` to try and rule out any inadvertent version-control problems. Still the `tap-next-release` doesn't work:
- Tapping `@spc` alone emits `shift` for a split second when `@spc` is released (this should emit " " instead)
- Pressing `@spc` then pressing `u` (just as an example) then releasing `@spc` emits "U" (this should emit " u" instead)
- I also have  `a (tap-hold 200 a lmet)` in `defalias` in my config. Pressing `@spc` then pressing `@a` then immediately releasing `@spc` emits "a" (it should emit " a" instead). This is especially strange because based on the previous bullet I would expect it to emit "A".
- Pressing `u` then pressing `@spc` then pressing `o` then releasing `u` emits "uO" (instead of "u")
- Pressing and holding `u` results in a sequence like "uuuuuuuuu" (because of my OS's key repeat - this is expected). While holding `u`, if I press `@spc` then tap `o` then the string of "u"s is interrupted with an "O", but then no more "u"s print, even though I haven't released `u` (is this expected behavior?)

It sounds like @obar didn't have any of these issues when trying my config. Hopefully `tap-next-release` works and these problems are on my end. I'll try and see what else could be going wrong over here.

**Comment by obar** [comment]
If I'm not mistaken, increasing the timeout for `tap-hold-next-release` to something very large will approximate `tap-next-release`, is that right?

I sometimes get the hold behavior when typing quickly using the `tap-hold-next-release`, and so I try again with a larger timeout. At this point that timeout is very large and is not related to the hold behavior showing up. If I change that key to a `tap-next-release` though, the problem doesn't exist.

I realize this isn't the best bug report but is it possible that the logic is different for the two buttons beyond just the timeout? Here's an example from quickly rolling `.,h` (did this a few times in a row, described behavior doesn't always happen) on a colemak layer, with `tap-hold-next-release` on `.` (`alt`) and `,` (`ctrl`) with a 2000ms timeout so it shouldn't affect anything.

```
--------------------------------------------------------------------------------
Received event: Press <.>
Running hooks
Block level set to: 1
Registering hook: 1525
Registering hook: 1526

--------------------------------------------------------------------------------
Received event: Press <,>
Running hooks
Registering hook: 1527
Registering hook: 1528
Storing event, current store: 
 - Press <,>

Cancelling hook: 1527
Emitting: Press <ralt>                <<< Shouldn't be happening
Registering hook: 1529
Unblocking input stream, rerunning:
Press <,>


--------------------------------------------------------------------------------
Rerunning event: Press <,>
Running hooks
Registering hook: 1530
Registering hook: 1531
Block level set to: 1
Registering hook: 1532
Registering hook: 1533

--------------------------------------------------------------------------------
Received event: Press <m>
Running hooks
Registering hook: 1534
Registering hook: 1535
Registering hook: 1536
Registering hook: 1537
Storing event, current store: 
 - Press <m>


--------------------------------------------------------------------------------
Received event: Release <.>
Running hooks
Registering hook: 1538
Registering hook: 1539
Emitting: Release <ralt>              <<< Expecting a tap of <.> here

--------------------------------------------------------------------------------
Received event: Release <,>
Running hooks
Emitting: Press <,>
Emitting: Release <,>
Unblocking input stream, rerunning:
Press <m>


--------------------------------------------------------------------------------
Rerunning event: Press <m>
Running hooks
Emitting: Press <h>
Registering hook: 1540

--------------------------------------------------------------------------------
Received event: Release <m>
Running hooks
Emitting: Release <h>

--------------------------------------------------------------------------------
```

**Comment by thoelze1** [comment]
I figured out the problem - it's Mac specific and the fix should be quick. For now, I hacked in a temporary fix and the tap-next-release works. Woohoo! I've been looking forward to this moment for months! So happy to have it implemented. Anyways, it's late here so I'll clean it up tomorrow.

Thanks @obar @david-janssen!

**Comment by david-janssen** [comment]
@obar : Very interesting. That indeed shouldn't be happening. Don't you just love bugs that trigger only occasionally? :-). I'll have a look at it sometime soon.

@thoelze1 : A bug that is only Mac-specific? Do explain! I was hoping that all of the KMonad internals would be OS independent. Does it have something to do with how the keysink and keysource are working? Anyways: looking forward to your fix and explanation.

Cheerio

**Comment by thoelze1** [comment]
@david-janssen I was hoping that too! My fix should address that. First I'll explain the issue:

On Mac, each key press (or key release) reported to us by the OS is accompanied by at least 3 additional key events that do not correspond to physical keys but rather store information about the keyboard state: which keys are currently pressed, proximity to a rollover error, etc. Here's a quick example:

Pressing <kbd>J</kbd> results in 4 key events (shown in the table) where

- Keycode `0xffff` is "reserved", its value denotes information about which keys are currently down or have changed state
- Keycode `0x1` is "rollover error"
- Keycode`0xd` corresponds to <kbd>J</kbd>

|Keycode        |Value          |Meaning   |
|--------------:|--------------:|:----------------------------------------------|
|`0xffff`       |`0xd`          |<kbd>J</kbd> was just pressed                  |
|`0x1`       |`0x0`          |There's no rollover error                      |
|`0xd`       |`0x1`          |Key down event for <kbd>J</kbd>                |
|`0xffff`       |`0xd`          |Currently pressed: <kbd>J</kbd>                |

Releasing <kbd>J</kbd> then results in 4 more key events:

|Keycode        |Value          |Meaning   |
|--------------:|--------------:|:----------------------------------------------|
|`0xffff`       |`0x0`          |A key was just released                  |
|`0x1`       |`0x0`          |There's no rollover error                      |
|`0xd`       |`0x0`          |Key up event for <kbd>J</kbd>                |
|`0xffff`       |`0x0`          |Currently pressed:                |

When I implemented the C code that listens for key events and repackages them for the Haskell code, I decided to treat these additional key events just like key down events and key up events: the C code simply passes them on to the Haskell code. I decided to do this to minimize the complexity of the C code, so that all processing could occur in Haskell. 

But there's a problem here. The "value" is inverted in the C code and passed to Haskell as the `Switch` value of a `KeyEvent`: 0 corresponds to `Press` and 1 corresponds to `Release`. Therefore that last key event when <kbd>J</kbd> is pressed (last line of the first table) is passed to Haskell as a key press event for `KeyReserved`. Then, when <kbd>J</kbd> is released, the first key event to be passed to Haskell (first line of the second table) is a key release event for `KeyReserved`. This always occurs before the key release event of <kbd>J</kbd>. Therefore tapping <kbd>J</kbd> (or any other key) always comes with a shorter tap of `KeyReserved` while <kbd>J</kbd> is down. This explains the behavior I reported in my comment above. It has probably affected other features of kmonad before `tap-next-release` but I only use `tap-hold` which seems to be unaffected.

The solution is to have the C code filter out key events that do not correspond to physical key presses/releases. This makes sense because:
- kmonad modifies a stream of key presses/releases; the additional key events do not represent key presses/releases
- kmonad reports modified key presses via a virtual keyboard. The state information of the physical keyboard should not be passed through kmonad to the virtual keyboard because it is not necessarily consistent with the keypresses as output by kmonad, and the virtual keyboard is not subject to rollover limitations

I will make a PR on testing with changes to the Mac code.

**Comment by david-janssen** [comment]
Aaahhh, that explanation makes a lot of sense. Good to hear that an OS specific issue was breaking down at the outer edge of the Haskell code. 

I see you already submitted a fix (approved), but I just wanted to mention that I actually really like the idea of keeping the C-code as simple as possible. To that extent it is possible to put most of the complexity in the Haskell code. The Linux key-IO is an example of this. The C-code basically just does the ioctl grab. Then haskell reads all the raw events from the file descriptor (equivalent to using the Mac c-code to dump every event into Haskell), and then Haskell has a function `fromLinuxKeyEvent` that only parses valid key events.

**Comment by thoelze1** [comment]
Okay, that makes sense. I'll have a look at `fromLinuxKeyEvent`...

**Comment by david-janssen** [comment]
If you'd like, however, if your code works it works, no worries

**Comment by david-janssen** [comment]
@obar Are you still experiencing issues with the tap-hold-release-next and tap-release-next buttons? If so, I'll have a deeper look, if not, then I'll close this issue soon.

**Comment by thoelze1** [comment]
For what it's worth, `tap-next-release` is working properly for me (I will reiterate that it provides the behavior I've been imagining for months and it feels wonderful to have my keyboard just work :) thanks everyone for your efforts).  That being said I have not tested `tap-hold-next-release`.

**Comment by david-janssen** [comment]
Great to hear that. I am currently writing a lot of documentation so that we can switch all of this refactor progress to master. Will be getting to 'tap-next-release' style stuff very soon, so I'll give it another test. If I can't reproduce @obar's issue, I will just close this for now.

**Comment by david-janssen** [comment]
Just dropping int to say: @obar was completely right. Something is fishy.

**Comment by david-janssen** [comment]
Hehe, found it :-).

It was super deep in KMonad and had nothing to do with the button itself. Turns out there was a problem in the function I used to calculate time-differences. It worked fine until we jumped a second. The logic itself was also fine except.... forgot about unsigned-integer mathematics. D'oh!

Fixed in 7decbf00c727097cbbc82d6bb0bcf435723e4073

Closing this now, because I think everything works.

**Comment by obar** [comment]
Wow, that's a tricky problem to find. How did you do it? Impressive bug fix! at this rate I'll have to set up a macro key to write: thanks David!

**Comment by david-janssen** [comment]
Trace statements, trace statements everywhere!! :-)

**Comment by molleweide** [comment]
tap-next-release is amazing
